secretariat/tests/test_models.py

```diff
@@ -32,3 +35,46 @@ def test_organisation_members(self):
             self.assertIsInstance(
                 user, User, "objects of organisation.members should be Users"
             )
+
+    def test_membership_cannot_end_before_beginning(self):
+        wrong_end_date = date.today() - timedelta(days=100)
+        membership = MembershipFactory(start_date=date.today(), end_date=wrong_end_date)
+        self.assertFalse(membership.start_date <= membership.end_date)
+        with self.assertRaises(ValidationError):
+            membership.full_clean()
+            membership.save()
```

Comment:
Je passe par `full_clean` pour lever la `ValidationError`, un peu à l'image des formulaires. Je regarde s'il y a un moyen de faire ça directement dans `.save()`.

Reply:
Normalement les `constraints` devraient faire lever une exception également.

Reply:
En tout cas, je m'attendais à ça, et si ça ne suffit pas, alors #1jour1leçon pour moi

Reply:
Et bien figure-toi que `AssertionError: ValidationError not raised`. Ce que je trouve étonnant aussi et je m'en vais réellement enquêter sur le pourquoi du comment. J'ai déjà dit ça il y a une heure mais à la place j'ai choisi de constater-avec-effarement-que-j'ai-oublié-la-cup-sur-le-feu puis descendre-à-la-pharmacie-pour-acheter-une-troisième-cup-cette-année.

Reply:
En fait, dans le cadre d'un formulaire, `.save()` est appelé après qu'on ait fait tourner `is_valid()` qui effectivement passer par les contraintes.

Reply:
Y a un sacré coquinou qui disait sur stackoverflow : 

def save(self, *args, **kwargs):
        if not self.is_cleaned:
            self.clean() 
            
 qu'en penses-tu ?

Reply:
Spontanément, je n'aime pas trop l'idée de surcharger save(), mais peut-être que je n'ai pas encore tous les codes Django/Python.

Reply:
Bon, en fait : c'est écrit dans la doc : "Les contraintes sont vérifiées durant la [validation des modèles](https://docs.djangoproject.com/fr/4.2/ref/models/instances/#validating-objects)."

Je pense qu'on aura moins de surprises si on laisse les contraintes dans la validation des modèles et que l'on teste celle-ci. Donc on laisse comme ça et on ne touche pas à save().